test/opt/instruction_list_test.cpp

```diff
@@ -0,0 +1,103 @@
+// Copyright (c) 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <memory>
+#include <vector>
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+#include "opt/instruction.h"
+#include "opt/instruction_list.h"
+
+namespace {
+
+using Instruction = spvtools::ir::Instruction;
+using InstructionList = spvtools::ir::InstructionList;
+using ::testing::ContainerEq;
+using ::testing::ElementsAre;
+using InstructionListTest = ::testing::Test;
+
+class TestInstruction : public Instruction {
+ public:
+  TestInstruction() : Instruction() {
+    created_instructions_.push_back(this);
+    static int counter = 0;
+    data = ++counter;
+  }
+
+  ~TestInstruction() { deleted_instructions_.push_back(this); }
+
+  int data;
+  static std::vector<TestInstruction*> created_instructions_;
+  static std::vector<TestInstruction*> deleted_instructions_;
+};
+
+std::vector<TestInstruction*> TestInstruction::created_instructions_;
+std::vector<TestInstruction*> TestInstruction::deleted_instructions_;
+
+// Test that the destructor for InstructionList is calling the destructor
+// for every element that is in the list.
+TEST(InstructionListTest, Destructor) {
+  InstructionList* list = new InstructionList();
+  list->push_back(new Instruction());
+  list->push_back(new Instruction());
+  delete list;
+
+  EXPECT_THAT(TestInstruction::created_instructions_,
+              ContainerEq(TestInstruction::deleted_instructions_));
+}
+
+// Test the |InsertBefore| with a single instruction in the iterator class.
+// Need to make sure the elements are inserted in the correct order, and the
+// return value points to the correct location.
+TEST(InstructionListTest, InsertBefore1) {
+  InstructionList list;
+  std::vector<Instruction*> created_instructions;
+  for (int i = 0; i < 4; i++) {
+    std::unique_ptr<Instruction> inst(new Instruction());
+    created_instructions.push_back(inst.get());
+    auto new_element = list.end().InsertBefore(std::move(inst));
+    EXPECT_THAT(&*new_element, created_instructions.back());
+  }
+
+  std::vector<Instruction*> output;
+  for (auto& i : list) {
+    output.push_back(&i);
+  }
+  EXPECT_THAT(output, ContainerEq(created_instructions));
```

Comment:
This seems awkwardly limiting as you can have output only in the order they were created.
Perhaps comparing data_ and not pointers would be better, as you can then specify in which order the elements should come.

Reply:
I renamed created_instructions to make it clearer.  The important part is that the instruction appear in the list in the same order as they are inserted.  Using "data_" is ackward because that field does not exist in an instruction.  If I use TestInstructions, there is lots of extra casting that is needed to access "data_".

The other consideration is that I want to make sure the address remain stable while moving things around.  We will have other data structures like the def-use manager that will have pointers to instructions.  It is important that we have stable address to help avoid dangling pointers.  I'll add a comment about that.